lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/impl/ThreadPoolStage.java

```diff
@@ -174,24 +174,32 @@ public ThreadPoolStage(@Parameter(StageName.class) final String name,
   @SuppressWarnings("checkstyle:illegalcatch")
   public void onNext(final T value) {
     beforeOnNext();
-    executor.submit(new Runnable() {
-
-      @Override
-      public void run() {
-        try {
-          handler.onNext(value);
-          afterOnNext();
-        } catch (final Throwable t) {
-          if (errorHandler != null) {
-            errorHandler.onNext(t);
-          } else {
-            LOG.log(Level.SEVERE, name + " Exception from event handler", t);
-            throw t;
+    try {
+      executor.submit(new Runnable() {
+
+        @Override
+        public void run() {
+          try {
+            handler.onNext(value);
+          } catch (final Throwable t) {
+            if (errorHandler != null) {
+              errorHandler.onNext(t);
+            } else {
+              LOG.log(Level.SEVERE, name + " Exception from event handler", t);
+              throw t;
+            }
+          } finally {
+            afterOnNext();
```

Comment:
It's a `protected` method in `AbstractEStage` that increments the `outMeter`. It should be called on exit of a handler.
Sorry, but where are you suggesting to do the bookkeeping?